### PR TITLE
azure: use a custom event loop for authentication

### LINKED ISSUE #5556: Issue with fetch from azure remote "no current event loop in thread"
# Bug Report

I'm getting this error
```
ERROR: failed to download 'azure://container/***' to '.dvc\cache\***' - There is no current event loop in thread 'ThreadPoolExecutor-127_0
...
ERROR: failed to download 'azure://container/***' to '.dvc\cache\***' - There is no current event loop in thread 'ThreadPoolExecutor-127_14
ERROR: failed to fetch data from the cloud - 53 files failed to download
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x000002159F326E88
```
For a number of files (not all as far I can tell) when running 
`dvc fetch [some_files]` on Azure devops windows CI machine with SPA auth.

DVC v2.0.3, everything seems to run fine on my local windows machine

This still fails if I set the number of jobs to 1

This is working fine with DVC v1.11.10

### PR DESCRIPTION
**Pull request by isidentical**
Resolves #5556

### CODE REVIEW

**Comment by isidentical** [comment]
@friggog can you try this patch to confirm whether it works for you or not? I've tested it by myself, and didn't hit any issues

**Comment by friggog** [comment]
Unfortunately not fixed:
```
DVC version: 2.0.18+59dbea
---------------------------------
Platform: Python 3.7.10 on Windows-10-10.0.19041-SP0
Supports: azure, http, https
Cache types: hardlink
Cache directory: NTFS on C:\
Caches: local
Remotes: azure
Workspace directory: NTFS on C:\
Repo: dvc (subdir), git
```

Same error as before when running dvc pull/fetch

```
Traceback (most recent call last):
  File "...lib\site-packages\dvc\remote\base.py", line 35, in wrapper
    func(from_info, to_info, *args, **kwargs)
  File "...lib\site-packages\dvc\fs\base.py", line 284, in download
    return self._download_file(from_info, to_info, name, no_progress_bar,)
  File "...lib\site-packages\dvc\fs\base.py", line 341, in _download_file
    from_info, tmp_file, name=name, no_progress_bar=no_progress_bar
  File "...lib\site-packages\dvc\fs\fsspec_wrapper.py", line 150, in _download
    with self.open(from_info, "rb") as fobj:
  File "...lib\site-packages\dvc\fs\fsspec_wrapper.py", line 86, in open
    return self.fs.open(self._with_bucket(path_info), mode=mode)
  File "...lib\site-packages\fsspec\spec.py", line 946, in open
    **kwargs,
  File "...lib\site-packages\adlfs\spec.py", line 1551, in _open
    **kwargs,
  File "...lib\site-packages\adlfs\spec.py", line 1615, in __init__
    fs.service_client.get_container_client(self.container_name)
  File "...lib\site-packages\azure\storage\blob\aio\_blob_service_client_async.py", line 591, in get_container_client
    key_resolver_function=self.key_resolver_function, loop=self._loop)
  File "...lib\site-packages\azure\storage\blob\aio\_container_client_async.py", line 118, in __init__
    **kwargs)
  File "...lib\site-packages\azure\storage\blob\_container_client.py", line 149, in __init__
    super(ContainerClient, self).__init__(parsed_url, service='blob', credential=credential, **kwargs)
  File "...lib\site-packages\azure\storage\blob\_shared\base_client.py", line 114, in __init__
    self._config, self._pipeline = self._create_pipeline(self.credential, storage_sdk=service, **kwargs)
  File "...lib\site-packages\azure\storage\blob\_shared\base_client_async.py", line 71, in _create_pipeline
    self._credential_policy = AsyncBearerTokenCredentialPolicy(credential, STORAGE_OAUTH_SCOPE)
  File "...lib\site-packages\azure\core\pipeline\policies\_authentication_async.py", line 24, in __init__
    self._lock = asyncio.Lock()
  File "...lib\asyncio\locks.py", line 161, in __init__
    self._loop = events.get_event_loop()
  File "...lib\asyncio\events.py", line 644, in get_event_loop
    % threading.current_thread().name)
RuntimeError: There is no current event loop in thread 'ThreadPoolExecutor-10_0'.
```

**Comment by isidentical** [comment]
> I think the issue arises when using the filesystem from within a ThreadPoolExecutor (e.g. here and here) rather than when initializing it

Generally, the methods of adlfs are always wrapped with an event loop, though there are 2 exceptions that I can see. One was the initialization, and now by looking at your traceback I also noticed that the `open()` calls are implemented separately and not wrapped with event loops. Can you try again with the latest version of this PR?

**Comment by friggog** [comment]
@isidentical seems to be working locally using DefaultAzureCredential! I'll check with SPA later, but seems like this resolves the issue 🎉

**Comment by isidentical** [comment]
Thanks for trying it out! We can probably wait for the SPA test, and then merge it.

**Comment by friggog** [comment]
Also works using SPA - I get these warnings after the command completes though:
```
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f9a4869db90>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f9a48670510>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f9a48616050>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f9a486097d0>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f9a48619c10>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f9a485cb7d0>
```
This didn't happen locally interestingly

**Comment by isidentical** [comment]
This is probably related to adlfs (unlikely) or the azure's python bindings (likely), where they don't close something somewhere. Honestly I think that we could ignore them just like we ignore normal asyncio warnings, since we are seeing those when they are getting deallocated so that means that is no leak or something like that.

File: dvc/fs/azure.py [hunk 1]
```diff
@@ -19,6 +21,34 @@
 )
 
 
+@contextmanager
+def _temp_event_loop():
```
[the remark this answers is not in the document]
Comment (isidentical) [reply dvc/fs/azure.py:25]: Ref: https://github.com/iterative/dvc/issues/5556#issuecomment-827537013

File: tests/unit/remote/test_azure.py [hunk 2]
```diff
@@ -1,4 +1,9 @@
-from dvc.fs.azure import AzureFileSystem
+import asyncio
```
Comment (efiop) [inline tests/unit/remote/test_azure.py:1]: Let's move this file to tests/unit/fs/test_azure.py while we are at it

File: dvc/fs/azure.py [hunk 3]
```diff
@@ -133,3 +164,9 @@ def fs(self):
             ) from e
 
         return file_system
+
+    def open(
+        self, path_info, mode="r", **kwargs
+    ):  # pylint: disable=arguments-differ
+        with _temp_event_loop():
```
Comment (efiop) [inline dvc/fs/azure.py:171]: Could you elaborate why we need another temp event loop here again, please?

Comment (isidentical) [reply dvc/fs/azure.py:171]: See: https://github.com/iterative/dvc/pull/5958#issuecomment-832521073

Comment (efiop) [reply dvc/fs/azure.py:171]: Sorry, blinded that one :facepalm: So I guess we are waiting for updates on https://github.com/dask/adlfs/issues/218 and then will maybe remove it. Gotcha.

**Review by efiop — APPROVED** [review]

**Pull request merged**